### PR TITLE
Add support for ganon

### LINKED ISSUE #108: [Feature]  Add support for ganon 'tre' report
### Checklist

- [X] There are [no similar issues or pull requests](https://github.com/taxprofiler/taxpasta/issues) for this yet.

### Problem

taxprofiler should support ganon output

### Solution

Support 

https://pirovc.github.io/ganon/outputfiles/#ganon-report

and/or

https://pirovc.github.io/ganon/outputfiles/#ganon-classify 

(the output is the same)

### Alternatives

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by jfy133**
Adds support for ganon

* [x] fix #108 
* [x] description of feature/fix
* [ ] tests added/passed
* [x] add an entry to the [changelog](../CHANGELOG.md)

### CODE REVIEW

**Comment by jfy133** [comment]
@Midnighter the error is as follows:

```
==================================== ERRORS ====================================
_ ERROR collecting tests/unit/infrastructure/application/ganon/test_ganon_profile.py _
tests/unit/infrastructure/application/ganon/test_ganon_profile.py:178: in <module>
    pd.DataFrame(
E   TypeError: DataFrame.__init__() got an unexpected keyword argument 'marks'
```

As far as we can tell (I think), the `test_<xxx>_profile.py` file is the same as centrifuge - so we are not really sure what is causing that.


Also we have a potential issue (don't know the behaviour yet of taxpasta until the tests are properly running), but the taxonomic ID for unclassified in `ganon` is `-`... do you have any suggestions hwo to deal with it (the other cases we saw were either NAs which works with pandas or just different integers)

File: .vscode/settings.json
Comment (Midnighter) [inline .vscode/settings.json:1]: Seems not a very meaningful change 😛

File: CHANGELOG.md [hunk 1]
```diff
@@ -12,6 +12,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 -   Provided curl commands in the intro to quickly fetch example files to try out (#97).
 -   Added new tutorial on merging across classifiers, and warnings why taxpasta
     does not currently do this natively (#98).
+-   Added classifier support for ganon report files
```
Comment (Midnighter) [inline CHANGELOG.md:15]:
```suggestion
-   Added classifier support for ganon report files (#109).
```

File: src/taxpasta/infrastructure/application/__init__.py [hunk 2]
```diff
@@ -61,6 +61,11 @@
     MotusProfileReader,
     MotusProfileStandardisationService,
 )
+from .ganon import (
+    GanonProfile,
+    GanonProfileReader,
+    GanonProfileStandardisationService,
+)
```
Comment (Midnighter) [inline src/taxpasta/infrastructure/application/__init__.py:68]: I've been trying to keep the profilers in alphabetical order 😉

File: src/taxpasta/infrastructure/application/application_service_registry.py [hunk 3]
```diff
@@ -79,7 +79,11 @@ def profile_reader(cls, profiler: SupportedProfiler) -> Type[ProfileReader]:
         elif profiler is SupportedProfiler.motus:
             from .motus import MotusProfileReader
 
-            return MotusProfileReader
+            return GanonProfileReader
```
Comment (Midnighter) [inline src/taxpasta/infrastructure/application/application_service_registry.py:82]:
```suggestion
            return MotusProfileReader
```

File: src/taxpasta/infrastructure/application/application_service_registry.py [hunk 4]
```diff
@@ -121,7 +125,12 @@ def profile_standardisation_service(
         elif profiler is SupportedProfiler.metaphlan:
             from .metaphlan import MetaphlanProfileStandardisationService
 
-            return MetaphlanProfileStandardisationService
+            return GanonProfileStandardisationService
```
Comment (Midnighter) [inline src/taxpasta/infrastructure/application/application_service_registry.py:128]:
```suggestion
            return MetaphlanProfileStandardisationService
```

File: src/taxpasta/infrastructure/application/ganon/ganon_profile.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        """Check that the percent of 'unclassified' and 'root' add up to a hundred."""
+        ## TODO: may need tweak tolerance amount, based on experience
+        ## ganon reports percentage to 5 decimal places
+        return profile.empty or bool(
+            np.isclose(
+                profile.loc[
+                    profile[cls.rank].isin(["unclassified", "root"]), cls.pc_cumulative
+                ].sum(),
+                100.0,
+                atol=0.001,
```
Comment (Midnighter) [inline src/taxpasta/infrastructure/application/ganon/ganon_profile.py:56]: This is much stricter than all the other profiles now. We mostly check with an absolute tolerance of 1% now.

File: src/taxpasta/infrastructure/application/ganon/ganon_profile.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+from taxpasta.infrastructure.helpers import BaseDataFrameModel
+
+
+class GanonProfile(BaseDataFrameModel):
+    """Define the expected ganon profile format."""
+
+    rank: Series[str] = pa.Field()
+    target: Series[int] = pa.Field(
+        ge=0
+    )  ## TODO: problem - unclassified gets assigned `-`
```
Comment (Midnighter) [inline src/taxpasta/infrastructure/application/ganon/ganon_profile.py:36]: My recommendation is that you define this as a `str` column, and then transform it in the standardisation service. Similarly to how it is done in `src/taxpasta/infrastructure/application/metaphlan/metaphlan_profile_standardisation_service.py`. Happy to take over that part.

File: src/taxpasta/infrastructure/application/ganon/ganon_profile_reader.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+                GanonProfile.rank,
+                GanonProfile.target,
+                GanonProfile.lineage,
+                GanonProfile.name,
+                GanonProfile.nr_unique,
+                GanonProfile.nr_shared,
+                GanonProfile.nr_children,
+                GanonProfile.nr_cumulative,
+                GanonProfile.pc_cumulative,
+            ],
```
Comment (Midnighter) [inline src/taxpasta/infrastructure/application/ganon/ganon_profile_reader.py:64]: So the ganon tool doesn't provide any column headers itself? How did you come up with the names for them, then?

**Review by Midnighter — CHANGES_REQUESTED** [review]
Very nice looking PR over all! Impressed to see that you even added the new ETL tests even, though they are not described in the [documentation](https://taxpasta.readthedocs.io/en/latest/contributing/supporting_new_profiler/) yet.

We can work together on finding a good solution for the target column.

File: src/taxpasta/infrastructure/application/__init__.py (hunk 2 above)
Comment (jfy133) [reply src/taxpasta/infrastructure/application/__init__.py:68]: Will fix everywhere! didn't notice that sorry!

File: src/taxpasta/infrastructure/application/ganon/ganon_profile.py (hunk 6 above)
Comment (jfy133) [reply src/taxpasta/infrastructure/application/ganon/ganon_profile.py:36]: Yes please do that would be nice of you <3, we couldn't work out exactly exactly how/where to do that bit (but @sofstam your intuition was right it seems!)

File: src/taxpasta/infrastructure/application/ganon/ganon_profile_reader.py (hunk 7 above)
Comment (jfy133) [reply src/taxpasta/infrastructure/application/ganon/ganon_profile_reader.py:64]: No ti does not, we took it from the output documentation: https://pirovc.github.io/ganon/outputfiles/#ganon-report

(replacing # with nr, and % with pc)

**Comment by jfy133** [comment]
> you even added the new ETL tests even

You can thank @sofstam for that! 

I've made the changes as requested @Midnighter , and  I added a line to the contributing docs to actually document (we ran out of time, then I forgot about them :facepalm: ).

File: src/taxpasta/infrastructure/application/ganon/ganon_profile.py (hunk 5 above)
Comment (Midnighter) [reply src/taxpasta/infrastructure/application/ganon/ganon_profile.py:56]: Ideally, we can confirm with a a range of real world profiles.

File: src/taxpasta/infrastructure/application/ganon/ganon_profile_reader.py (hunk 7 above)
Comment (Midnighter) [reply src/taxpasta/infrastructure/application/ganon/ganon_profile_reader.py:64]: Would you be very opposed to making them more expressive, like `nr` -> `number` and `pc` -> `percent`?

**Comment by Midnighter** [comment]
Thanks for the changes. I'm working on the standardisation now. The error that you saw

```
TypeError: DataFrame.__init__() got an unexpected keyword argument 'marks'
```

was caused by not using `pytest.param` in some cases. (Probably just a copy & paste mistake.)

File: src/taxpasta/infrastructure/application/ganon/ganon_profile_reader.py (hunk 7 above)
Comment (jfy133) [reply src/taxpasta/infrastructure/application/ganon/ganon_profile_reader.py:64]: No not at all! Go ahead

**Comment by jfy133** [comment]
[ganon_testdata.zip](https://github.com/taxprofiler/taxpasta/files/11865540/ganon_testdata.zip)

Mor etest data for you @Midnighter !

**Review by Midnighter — APPROVED** [review]

**Pull request merged**